### ISSUE TITLE
Inference missing obvious type from a `QuoteNode`

### ISSUE THREAD

**Issue by andyferris** [JuliaLang/julia#20801]
On master, I'm getting an odd message from `@code_warntype` where inference seems to have failed to have picked up  a type returned by another function (in this case, `Base.promote_op`). An example of the output that I see is this:

```julia
julia> include("recursive.jl")

julia> v = Vec{4,Float64}((1.0,2.0,3.0,4.0))
Vec{4,Float64}((1.0, 2.0, 3.0, 4.0))

julia> @code_warntype map(x -> 2x, v)
Variables:
  #self#::Base.#map
  f::##7#8
  m::Vec{4,Float64}
  T2::Any

Body:
  begin 
      T2::Any = $(QuoteNode(Float64)) # line 14:
      $(Expr(:inbounds, true))
      SSAValue(0) = (Main.map_1d)(f::##7#8, (Core.getfield)(Main.v, :data)::Any)::Tuple
      return $(Expr(:new, :($(QuoteNode(Vec{4,Float64}))), :((Base.convert)($(QuoteNode(NTuple{4,Float64})), SSAValue(0))::NTuple{4,Any})))
      $(Expr(:inbounds, :pop))
  end::Vec{4,Float64}
```

The problem is the first line in the body: `T2::Any = $(QuoteNode(Float64)) # line 14:`. Somehow the type of `T2` wasn't inferred.

You can find `recursive.jl` at [this gist](https://gist.github.com/andyferris/d7d35b98c654b788828c9c7ac944d307).

**Comment by yuyichao** [comment]
Note that there's no type instability caused by that, though it would be good to hide the unused `T2`.

**Comment by andyferris** [comment]
My understanding was that line of untyped code was causing bad codegen. I think @SimonDanisch mentioned seeing something similar (on slack).

**Comment by yuyichao** [comment]
Not for this case. Your error is at https://gist.github.com/andyferris/d7d35b98c654b788828c9c7ac944d307#file-recursive-jl-L14

**Comment by KristofferC** [comment]
Ref https://github.com/JuliaLang/julia/issues/21281

**Comment by KristofferC** [comment]
What more info need to be provided?

**Comment by yuyichao** [comment]
There are two issues here. One is a harmless but confusing warning for variables that are not used (the `T2`, i.e. #21281). The other is the claim that it actually causes bad code to be generated. The code in the first post did generate bad code but that's because an error in the code so the "more info" is

> My understanding was that line of untyped code was causing bad codegen. I think @SimonDanisch mentioned seeing something similar (on slack).

**Comment by KristofferC** [comment]
This would be nice to have fixed. It makes finding type instabilities in largish function much harder.

**Comment by rleegates** [comment]
Perhaps this is related, can't seem to find the problem on why `T3` is not inferred correctly, see the line containing `T3::Type{T} = $(QuoteNode(BlockSparse.Block{Array{Float64,2}}))` below. This, however, is on v0.5.1.

The type `Block{T} <: AbstractDataBlock{T}` has only one field `data::T`, where `T` is usually a matrix.

I overloaded:
```julia
promote_type{T,T2}(::Type{Block{Matrix{T}}}, ::Type{Block{Matrix{T2}}}) = Block{Matrix{promote_type(T,T2)}}
```

Code for the sparse matrix multiply function is in this [gist](https://gist.github.com/rleegates/d99050d89a06c9c6c117935935a421f4) and was modified from `Base` as the function in `Base` needs `zero(Tv)`, which I can't provide.

`code_warntype` output:
```julia
julia> @code_warntype Htr*Kraw;
Variables:
  #self#::Base.#*
  A::SparseMatrixCSC{BlockSparse.Block{Array{Int64,2}},Int64}
  B::SparseMatrixCSC{BlockSparse.Block{Array{Float64,2}},Int64}
  #temp#@_4::Int64
  mA::Int64
  nA::Int64
  #temp#@_7::Int64
  mB::Int64
  nB::Int64
  colptrA::Array{Int64,1}
  rowvalA::Array{Int64,1}
  nzvalA::Array{BlockSparse.Block{Array{Int64,2}},1}
  colptrB::Array{Int64,1}
  rowvalB::Array{Int64,1}
  nzvalB::Array{BlockSparse.Block{Array{Float64,2}},1}
  nnzC::Any
  colptrC::Array{Int64,1}
  rowvalC::Array{Int64,1}
  T3::Type{T}
  nzvalC::Any
  ip::Int64
  xb::Array{Int64,1}
  x::Any
  Cunsorted::SparseMatrixCSC{Tv,Int64}
  C::SparseMatrixCSC{Tv,Int64}
  #temp#@_26::Int64
  i::Int64
  #temp#@_28::Int64
  vp::Int64
  #temp#@_30::Int64
  jp::Int64
  nzB::BlockSparse.Block{Array{Float64,2}}
  j::Int64
  #temp#@_34::Int64
  kp::Int64
  nzC::BlockSparse.Block{Array{Float64,2}}
  k::Int64

Body:
  begin 
      NewvarNode(:(colptrA::Array{Int64,1}))
      NewvarNode(:(rowvalA::Array{Int64,1}))
      NewvarNode(:(nzvalA::Array{BlockSparse.Block{Array{Int64,2}},1}))
      NewvarNode(:(colptrB::Array{Int64,1}))
      NewvarNode(:(rowvalB::Array{Int64,1}))
      NewvarNode(:(nzvalB::Array{BlockSparse.Block{Array{Float64,2}},1}))
      NewvarNode(:(nnzC::Any))
      NewvarNode(:(colptrC::Array{Int64,1}))
      NewvarNode(:(rowvalC::Array{Int64,1}))
      NewvarNode(:(T3::Type{T}))
      NewvarNode(:(nzvalC::Any))
      NewvarNode(:(ip::Int64))
      NewvarNode(:(xb::Array{Int64,1}))
      NewvarNode(:(x::Any))
      NewvarNode(:(Cunsorted::SparseMatrixCSC{Tv,Int64}))
      NewvarNode(:(C::SparseMatrixCSC{Tv,Int64}))
      SSAValue(29) = (Core.getfield)(A::SparseMatrixCSC{BlockSparse.Block{Array{Int64,2}},Int64},:m)::Int64
      SSAValue(30) = (Core.getfield)(A::SparseMatrixCSC{BlockSparse.Block{Array{Int64,2}},Int64},:n)::Int64
      #temp#@_4::Int64 = $(QuoteNode(1))
      SSAValue(31) = (Base.box)(Int64,(Base.add_int)(1,1))
      mA::Int64 = SSAValue(29)
      #temp#@_4::Int64 = SSAValue(31)
      SSAValue(32) = (Base.box)(Int64,(Base.add_int)(2,1))
      nA::Int64 = SSAValue(30)
      #temp#@_4::Int64 = SSAValue(32) # line 122:
      SSAValue(33) = (Core.getfield)(B::SparseMatrixCSC{BlockSparse.Block{Array{Float64,2}},Int64},:m)::Int64
      SSAValue(34) = (Core.getfield)(B::SparseMatrixCSC{BlockSparse.Block{Array{Float64,2}},Int64},:n)::Int64
      #temp#@_7::Int64 = $(QuoteNode(1))
      SSAValue(35) = (Base.box)(Int64,(Base.add_int)(1,1))
      mB::Int64 = SSAValue(33)
      #temp#@_7::Int64 = SSAValue(35)
      SSAValue(36) = (Base.box)(Int64,(Base.add_int)(2,1))
      nB::Int64 = SSAValue(34)
      #temp#@_7::Int64 = SSAValue(36) # line 123:
      SSAValue(6) = (nA::Int64 === mB::Int64)::Bool
      unless SSAValue(6) goto 40
      goto 42
      40: 
      (BlockSparse.throw)($(Expr(:new, :(Base.DimensionMismatch), "")))::Union{}
      42:  # line 125:
      colptrA::Array{Int64,1} = (Core.getfield)(A::SparseMatrixCSC{BlockSparse.Block{Array{Int64,2}},Int64},:colptr)::Array{Int64,1} # line 125:
      rowvalA::Array{Int64,1} = (Core.getfield)(A::SparseMatrixCSC{BlockSparse.Block{Array{Int64,2}},Int64},:rowval)::Array{Int64,1} # line 125:
      nzvalA::Array{BlockSparse.Block{Array{Int64,2}},1} = (Core.getfield)(A::SparseMatrixCSC{BlockSparse.Block{Array{Int64,2}},Int64},:nzval)::Array{BlockSparse.Block{Array{Int64,2}},1} # line 126:
      colptrB::Array{Int64,1} = (Core.getfield)(B::SparseMatrixCSC{BlockSparse.Block{Array{Float64,2}},Int64},:colptr)::Array{Int64,1} # line 126:
      rowvalB::Array{Int64,1} = (Core.getfield)(B::SparseMatrixCSC{BlockSparse.Block{Array{Float64,2}},Int64},:rowval)::Array{Int64,1} # line 126:
      nzvalB::Array{BlockSparse.Block{Array{Float64,2}},1} = (Core.getfield)(B::SparseMatrixCSC{BlockSparse.Block{Array{Float64,2}},Int64},:nzval)::Array{BlockSparse.Block{Array{Float64,2}},1} # line 128:
      SSAValue(17) = (Base.box)(Int64,(Base.mul_int)(mA::Int64,nB::Int64))
      SSAValue(16) = (Base.box)(Int64,(Base.add_int)((Base.arraylen)(nzvalA::Array{BlockSparse.Block{Array{Int64,2}},1})::Int64,(Base.arraylen)(nzvalB::Array{BlockSparse.Block{Array{Float64,2}},1})::Int64))
      nnzC::Any = (Base.select_value)((Base.slt_int)(SSAValue(16),SSAValue(17))::Bool,SSAValue(16),SSAValue(17))::Int64 # line 129:
      colptrC::Array{Int64,1} = (Core.ccall)(:jl_alloc_array_1d,(Core.apply_type)(Core.Array,Int64,1)::Type{Array{Int64,1}},(Core.svec)(Core.Any,Core.Int)::SimpleVector,Array{Int64,1},0,(Base.box)(Int64,(Base.add_int)(nB::Int64,1)),0)::Array{Int64,1} # line 130:
      rowvalC::Array{Int64,1} = (Core.ccall)(:jl_alloc_array_1d,(Core.apply_type)(Core.Array,Int64,1)::Type{Array{Int64,1}},(Core.svec)(Core.Any,Core.Int)::SimpleVector,Array{Int64,1},0,nnzC::Int64,0)::Array{Int64,1} # line 131:
      T3::Type{T} = $(QuoteNode(BlockSparse.Block{Array{Float64,2}})) # line 132:
      nzvalC::Any = ((Core.apply_type)(BlockSparse.Array,T3::Type{T})::Type{_<:Array{T,N}})(nnzC::Int64)::Any # line 134:
      $(Expr(:inbounds, true)) # line 135:
      ip::Int64 = 1 # line 136:
      xb::Array{Int64,1} = $(Expr(:invoke, LambdaInfo for fill!(::Array{Int64,1}, ::Int64), :(Base.fill!), :((Core.ccall)(:jl_alloc_array_1d,(Core.apply_type)(Core.Array,Int64,1)::Type{Array{Int64,1}},(Core.svec)(Core.Any,Core.Int)::SimpleVector,Array{Int64,1},0,mA,0)::Array{Int64,1}), 0)) # line 137:
      x::Any = ((Core.apply_type)(BlockSparse.Vector,(Core.apply_type)(BlockSparse.Queue,T3::Type{T})::Type{_<:BlockSparse.Queue{T}})::Type{_<:Array{_<:BlockSparse.Queue{T},1}})(mA::Int64)::Any # line 138:
      SSAValue(37) = (Base.select_value)((Base.sle_int)(1,nB::Int64)::Bool,nB::Int64,(Base.box)(Int64,(Base.sub_int)(1,1)))::Int64
      #temp#@_26::Int64 = 1
      78: 
      unless (Base.box)(Base.Bool,(Base.not_int)((#temp#@_26::Int64 === (Base.box)(Int64,(Base.add_int)(SSAValue(37),1)))::Bool)) goto 162
      SSAValue(38) = #temp#@_26::Int64
      SSAValue(39) = (Base.box)(Int64,(Base.add_int)(#temp#@_26::Int64,1))
      i::Int64 = SSAValue(38)
      #temp#@_26::Int64 = SSAValue(39) # line 139:
      unless (nnzC::Any < (Base.box)(Int64,(Base.sub_int)((Base.box)(Int64,(Base.add_int)(ip::Int64,mA::Int64)),1)))::Any goto 92 # line 140:
      (BlockSparse.resize!)(rowvalC::Array{Int64,1},(nnzC::Any + (BlockSparse.max)(nnzC::Any,mA::Int64)::Any)::Any)::Array{Int64,1} # line 141:
      (BlockSparse.resize!)(nzvalC::Any,(nnzC::Any + (BlockSparse.max)(nnzC::Any,mA::Int64)::Any)::Any)::Any # line 142:
      nnzC::Any = (BlockSparse.length)(nzvalC::Any)::Any
      92:  # line 144:
      (Base.arrayset)(colptrC::Array{Int64,1},ip::Int64,i::Int64)::Array{Int64,1} # line 145:
      SSAValue(22) = (Base.arrayref)(colptrB::Array{Int64,1},i::Int64)::Int64
      SSAValue(21) = (Base.box)(Int64,(Base.sub_int)((Base.arrayref)(colptrB::Array{Int64,1},(Base.box)(Int64,(Base.add_int)(i::Int64,1)))::Int64,1))
      SSAValue(40) = (Base.select_value)((Base.sle_int)(SSAValue(22),SSAValue(21))::Bool,SSAValue(21),(Base.box)(Int64,(Base.sub_int)(SSAValue(22),1)))::Int64
      #temp#@_30::Int64 = SSAValue(22)
      100: 
      unless (Base.box)(Base.Bool,(Base.not_int)((#temp#@_30::Int64 === (Base.box)(Int64,(Base.add_int)(SSAValue(40),1)))::Bool)) goto 143
      SSAValue(41) = #temp#@_30::Int64
      SSAValue(42) = (Base.box)(Int64,(Base.add_int)(#temp#@_30::Int64,1))
      jp::Int64 = SSAValue(41)
      #temp#@_30::Int64 = SSAValue(42) # line 146:
      nzB::BlockSparse.Block{Array{Float64,2}} = (Base.arrayref)(nzvalB::Array{BlockSparse.Block{Array{Float64,2}},1},jp::Int64)::BlockSparse.Block{Array{Float64,2}} # line 147:
      j::Int64 = (Base.arrayref)(rowvalB::Array{Int64,1},jp::Int64)::Int64 # line 148:
      SSAValue(24) = (Base.arrayref)(colptrA::Array{Int64,1},j::Int64)::Int64
      SSAValue(23) = (Base.box)(Int64,(Base.sub_int)((Base.arrayref)(colptrA::Array{Int64,1},(Base.box)(Int64,(Base.add_int)(j::Int64,1)))::Int64,1))
      SSAValue(43) = (Base.select_value)((Base.sle_int)(SSAValue(24),SSAValue(23))::Bool,SSAValue(23),(Base.box)(Int64,(Base.sub_int)(SSAValue(24),1)))::Int64
      #temp#@_34::Int64 = SSAValue(24)
      115: 
      unless (Base.box)(Base.Bool,(Base.not_int)((#temp#@_34::Int64 === (Base.box)(Int64,(Base.add_int)(SSAValue(43),1)))::Bool)) goto 141
      SSAValue(44) = #temp#@_34::Int64
      SSAValue(45) = (Base.box)(Int64,(Base.add_int)(#temp#@_34::Int64,1))
      kp::Int64 = SSAValue(44)
      #temp#@_34::Int64 = SSAValue(45) # line 149:
      nzC::BlockSparse.Block{Array{Float64,2}} = $(Expr(:invoke, LambdaInfo for *(::BlockSparse.Block{Array{Int64,2}}, ::BlockSparse.Block{Array{Float64,2}}), :(BlockSparse.*), :((Base.arrayref)(nzvalA,kp)::BlockSparse.Block{Array{Int64,2}}), :(nzB))) # line 150:
      k::Int64 = (Base.arrayref)(rowvalA::Array{Int64,1},kp::Int64)::Int64 # line 151:
      unless (Base.box)(Base.Bool,(Base.not_int)(((Base.arrayref)(xb::Array{Int64,1},k::Int64)::Int64 === i::Int64)::Bool)) goto 136 # line 152:
      (Base.arrayset)(rowvalC::Array{Int64,1},k::Int64,ip::Int64)::Array{Int64,1} # line 153:
      ip::Int64 = (Base.box)(Int64,(Base.add_int)(ip::Int64,1)) # line 154:
      (Base.arrayset)(xb::Array{Int64,1},i::Int64,k::Int64)::Array{Int64,1} # line 155:
      (BlockSparse.setindex!)(x::Any,nzC::BlockSparse.Block{Array{Float64,2}},k::Int64)::Any
      goto 139
      136:  # line 157:
      (BlockSparse.add!)((BlockSparse.getindex)(x::Any,k::Int64)::Any,nzC::BlockSparse.Block{Array{Float64,2}})::Array{BlockSparse.Block{Array{Float64,2}},1}
      139: 
      goto 115
      141: 
      goto 100
      143:  # line 161:
      SSAValue(26) = (Base.arrayref)(colptrC::Array{Int64,1},i::Int64)::Int64
      SSAValue(25) = (Base.box)(Int64,(Base.sub_int)(ip::Int64,1))
      SSAValue(46) = (Base.select_value)((Base.sle_int)(SSAValue(26),SSAValue(25))::Bool,SSAValue(25),(Base.box)(Int64,(Base.sub_int)(SSAValue(26),1)))::Int64
      #temp#@_28::Int64 = SSAValue(26)
      149: 
      unless (Base.box)(Base.Bool,(Base.not_int)((#temp#@_28::Int64 === (Base.box)(Int64,(Base.add_int)(SSAValue(46),1)))::Bool)) goto 160
      SSAValue(47) = #temp#@_28::Int64
      SSAValue(48) = (Base.box)(Int64,(Base.add_int)(#temp#@_28::Int64,1))
      vp::Int64 = SSAValue(47)
      #temp#@_28::Int64 = SSAValue(48) # line 162:
      SSAValue(15) = (BlockSparse.reduce)((BlockSparse.getindex)(x::Any,(Base.arrayref)(rowvalC::Array{Int64,1},vp::Int64)::Int64)::Any)::Any
      (BlockSparse.setindex!)(nzvalC::Any,SSAValue(15),vp::Int64)::Any
      158: 
      goto 149
      160: 
      goto 78
      162:  # line 165:
      (Base.arrayset)(colptrC::Array{Int64,1},ip::Int64,(Base.box)(Int64,(Base.add_int)(nB::Int64,1)))::Array{Int64,1}
      $(Expr(:inbounds, :pop)) # line 168:
      SSAValue(28) = (Base.arrayref)(colptrC::Array{Int64,1},(Base.arraylen)(colptrC::Array{Int64,1})::Int64)::Int64
      SSAValue(27) = (Base.arraylen)(rowvalC::Array{Int64,1})::Int64
      $(Expr(:invoke, LambdaInfo for deleteat!(::Array{Int64,1}, ::UnitRange{Int64}), :(BlockSparse.deleteat!), :(rowvalC), :($(Expr(:new, UnitRange{Int64}, SSAValue(28), :((Base.select_value)((Base.sle_int)(SSAValue(28),SSAValue(27))::Bool,SSAValue(27),(Base.box)(Int64,(Base.sub_int)(SSAValue(28),1)))::Int64)))))) # line 169:
      (BlockSparse.deleteat!)(nzvalC::Any,(BlockSparse.colon)((Base.arrayref)(colptrC::Array{Int64,1},(Base.arraylen)(colptrC::Array{Int64,1})::Int64)::Int64,(BlockSparse.length)(nzvalC::Any)::Any)::Any)::Any # line 172:
      Cunsorted::SparseMatrixCSC{Tv,Int64} = (BlockSparse.SparseMatrixCSC)(mA::Int64,nB::Int64,colptrC::Array{Int64,1},rowvalC::Array{Int64,1},nzvalC::Any)::SparseMatrixCSC{Tv,Int64} # line 173:
      C::SparseMatrixCSC{Tv,Int64} = (BlockSparse.sortSparseMatrixCSC!)(Cunsorted::SparseMatrixCSC{Tv,Int64})::SparseMatrixCSC{Tv,Int64} # line 174:
      return C::SparseMatrixCSC{Tv,Int64}
  end::SparseMatrixCSC{Tv,Int64}

julia> promote_type(Block{Matrix{Int}},Block{Matrix{Float64}})
BlockSparse.Block{Array{Float64,2}}
```

**Comment by fredrikekre** [comment]
Fixed by #23280?

**Comment by KristofferC** [comment]
There is still the comment

> My understanding was that line of untyped code was causing bad codegen. I think @SimonDanisch mentioned seeing something similar (on slack).

Please open a new issue for it if that is still happening @SimonDanisch
